src/libraries/System.Private.CoreLib/src/System/IO/BufferedFileStreamStrategy.cs

```diff
@@ -0,0 +1,1079 @@
+            }
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            try
+            {
+                if (!_strategy.IsClosed)
+                {
+                    try
+                    {
+                        await FlushAsync().ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        await _strategy.DisposeAsync().ConfigureAwait(false);
+                    }
+                }
+            }
+            finally
+            {
+                // Don't set the buffer to null, to avoid a NullReferenceException
+                // when users have a race condition in their code (i.e. they call
+                // FileStream.Close when calling another method on FileStream like Read).
+
+                _writePos = 0; // WriteByte hot path relies on this
+            }
+        }
+
+        internal override void DisposeInternal(bool disposing) => Dispose(disposing);
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing && !_strategy.IsClosed)
+                {
+                    try
+                    {
+                        Flush();
+                    }
+                    finally
+                    {
+                        _strategy.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                // Don't set the buffer to null, to avoid a NullReferenceException
+                // when users have a race condition in their code (i.e. they call
+                // FileStream.Close when calling another method on FileStream like Read).
+
+                // Call base.Dispose(bool) to cleanup async IO resources
+                base.Dispose(disposing);
+
+                _writePos = 0;
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            AssertBufferArguments(buffer, offset, count);
+
+            return ReadSpan(new Span<byte>(buffer, offset, count), new ArraySegment<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> destination)
+        {
+            EnsureNotClosed();
+
+            return ReadSpan(destination, default);
+        }
+
+        private int ReadSpan(Span<byte> destination, ArraySegment<byte> arraySegment)
+        {
+            Debug.Assert((_readPos == 0 && _readLen == 0 && _writePos >= 0) || (_writePos == 0 && _readPos <= _readLen),
+                "We're either reading or writing, but not both.");
+
+            bool isBlocked = false;
+            int n = _readLen - _readPos;
+            // if the read buffer is empty, read into either user's array or our
+            // buffer, depending on number of bytes user asked for and buffer size.
+            if (n == 0)
+            {
+                EnsureCanRead();
```

Comment:
The same for many other places in this type. Consider removing some validations in `BufferedFileStreamStrategy`, such as `EnsureCanRead/Write` since they are already performed when calling the underlying `_strategy` methods.

Reply:
I've tried to keep them minimum, but in some cases they are mandatory, like here where `FileStream.Read(Span)` does not ensure that `CanRead == true`:

https://github.com/dotnet/runtime/blob/ac5c33defa6a9983e85ce8bf0be352cb09f5b8eb/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L226